src/start_server.ts

```diff
@@ -44,33 +53,59 @@ export interface ServerOptions {
 
   /** The package name to use for the root directory **/
   packageName?: string;
+
+  /** The HTTP protocol to use */
+  protocol?: string;
+
+  /** Path to TLS service key for HTTPS */
+  keyPath?: string;
+
+  /** Path to TLS certificate for HTTPS */
+  certPath?: string;
+
+  /** Path to H2 push-manifest file */
+  pushManifestPath?: string;
 }
 
-function applyDefaultOptions(options: ServerOptions): ServerOptions {
+async function applyDefaultOptions(options: ServerOptions): Promise<ServerOptions> {
   const withDefaults = Object.assign({}, options);
   Object.assign(withDefaults, {
-    port: options.port || 8080,
-    hostname: options.hostname || "localhost",
+    port: await nextOpenPort(options.port),
+    hostname: options.hostname || 'localhost',
     root: path.resolve(options.root || '.'),
+    certPath: options.certPath || 'cert.pem',
+    keyPath: options.keyPath || 'key.pem',
   });
   return withDefaults;
 }
 
 /**
- * @return {Promise} A Promise that completes when the server has started.
+ * If port unspecified/negative, finds an open port on localhost
+ * @param {number} port
+ * @returns {Promise<number>} Promise of open port
  */
-export function startServer(options: ServerOptions): Promise<http.Server> {
-  return new Promise((resolve, reject) => {
-    options = options || {};
-    if (options.port) {
-      resolve(options);
-    } else {
-      findPort(8080, 8180, (ports) => {
-        options.port = ports[0];
-        resolve(options);
+async function nextOpenPort(port: number): Promise<number> {
+  if (!port || port < 0) {
+    port = await new Promise<number>(resolve => {
+      findPort(8080, 8180, (ports: number[]) => {
+        resolve(ports[0]);
```

Comment:
Interestingly, I think `find-port` has a bug because it still resolves with a port number if none are available. The redundant error checking in `startWithPort()` catches the issue. The [`getport` module](https://www.npmjs.com/package/getport) (which has the same API as `find-port`) correctly returns an error in a callback. We could switch if you prefer.